src/lexer.cpp

```diff
@@ -102,244 +148,84 @@ std::pair<bool, TokenTag> Lexer::is_keyword(InternedString const& str) {
 	return {false, TokenTag::END};
 }
 
-void Lexer::consume_token() {
-
-	switch (current_char()) {
-
-	case '+':
-		switch (next_char()) {
-		case '+':
-			push_token(TokenTag::INCREMENT, 2);
-			break;
-		case '=':
-			push_token(TokenTag::ADD_TO, 2);
-			break;
-		default:
-			push_token(TokenTag::ADD, 1);
-		}
-		break;
-
-	case '-':
-		switch (next_char()) {
-		case '-':
-			push_token(TokenTag::DECREMENT, 2);
-			break;
-		case '=':
-			push_token(TokenTag::SUB_TO, 2);
-			break;
-		default:
-			push_token(TokenTag::SUB, 1);
-		}
-		break;
-
-	case '*':
-		switch (next_char()) {
-		case '=':
-			push_token(TokenTag::MUL_TO, 2);
-			break;
-		default:
-			push_token(TokenTag::MUL, 1);
-		}
-		break;
+bool Lexer::consume_string() {
+	if (current_char() != '"')
+		return false;
 
-	case '/':
-		switch (next_char()) {
-		case '/':
-			assert(consume_comment());
-			break;
-		case '=':
-			push_token(TokenTag::DIV_TO, 2);
-			break;
-		default:
-			push_token(TokenTag::DIV, 1);
-		}
-		break;
+	m_source_index += 1;
+	m_current_column += 1;
 
-	case '&':
-		switch (next_char()) {
-		case '&':
-			push_token(TokenTag::LOGIC_AND, 2);
-			break;
-		case '=':
-			push_token(TokenTag::AND_TO, 2);
-			break;
-		default:
-			push_token(TokenTag::AND, 1);
-		}
-		break;
+	int i0 = m_source_index;
+	int c0 = m_current_column;
+	int l0 = m_current_line;
 
-	case '|':
-		switch (next_char()) {
-		case '|':
-			push_token(TokenTag::LOGIC_IOR, 2);
-			break;
-		case '>':
-			push_token(TokenTag::PIZZA, 2);
-			break;
-		case '=':
-			push_token(TokenTag::IOR_TO, 2);
-			break;
-		default:
-			push_token(TokenTag::IOR, 1);
+	// TODO: support escape sequences
+	size_t len = 0;
+	while ((not done()) && current_char() != '"') {
+		len += 1;
+		m_source_index += 1;
+		if (current_char() == '\n') {
+			m_current_line += 1;
+			m_current_column = 0;
+		} else {
+			m_current_column += 1;
 		}
-		break;
+	}
 
-	case '^':
-		switch (next_char()) {
-		/* case '^': push_token(TokenTag::LOGIC_XOR, 2); break; */
-		case '=':
-			push_token(TokenTag::XOR_TO, 2);
-			break;
-		default:
-			push_token(TokenTag::XOR, 1);
-		}
-		break;
+	if (current_char() != '"') {
+		// TODO: report unmatched quote
+		assert(0);
+	}
 
-	case ':':
-		switch (next_char()) {
-		case '=':
-			push_token(TokenTag::DECLARE_ASSIGN, 2);
-			break;
-		default:
-			push_token(TokenTag::DECLARE, 1);
-		}
-		break;
+	char const* base_ptr = m_source.data() + i0;
+	m_tokens.push_back(
+			{TokenTag::STRING, InternedString {base_ptr, len}, l0, c0, m_current_line, m_current_column});
```

Comment:
This seems indented weirdly.

Reply:
Yeah I just cut and pasted without worrying about it